components/psst/browser/content/psst_tab_helper.cc

```diff
@@ -0,0 +1,136 @@
+                                              const int32_t world_id) {
+  if (contents->GetBrowserContext()->IsOffTheRecord() ||
+      !base::FeatureList::IsEnabled(psst::features::kBravePsst)) {
+    return;
+  }
+
+  psst::PsstTabHelper::CreateForWebContents(contents, world_id);
+}
+
+PsstTabHelper::PsstTabHelper(content::WebContents* web_contents,
+                             const int32_t world_id)
+    : WebContentsObserver(web_contents),
+      content::WebContentsUserData<PsstTabHelper>(*web_contents),
+      world_id_(world_id),
+      psst_rule_service_(PsstRuleService::GetInstance()) {
+  DCHECK(psst_rule_service_);
+}
+
+PsstTabHelper::~PsstTabHelper() = default;
+
+void PsstTabHelper::OnTestScriptResult(
+    std::string policy_script,
+    const content::GlobalRenderFrameHostId render_frame_host_id,
+    const int32_t world_id,
+    base::Value value) {
+  if (value.GetIfBool().value_or(false)) {
+    InsertScriptInPage(render_frame_host_id, world_id, policy_script,
+                       base::DoNothing());
+  }
+}
+
+void PsstTabHelper::InsertTestScript(
+    const content::GlobalRenderFrameHostId render_frame_host_id,
+    const int32_t world_id,
+    MatchedRule rule) {
+  CHECK(render_frame_host_id);
+  InsertScriptInPage(
+      render_frame_host_id, world_id, rule.test_script,
+      base::BindOnce(&PsstTabHelper::OnTestScriptResult,
+                     weak_factory_.GetWeakPtr(), rule.policy_script,
+                     render_frame_host_id, world_id));
+}
+
+void PsstTabHelper::InsertScriptInPage(
+    const content::GlobalRenderFrameHostId render_frame_host_id,
+    const int32_t world_id,
+    const std::string& script,
+    content::RenderFrameHost::JavaScriptResultCallback cb) {
+  content::RenderFrameHost* render_frame_host =
+      content::RenderFrameHost::FromID(render_frame_host_id);
+
+  // Check if render_frame_host is still valid and if starting rfh is the same.
+  if (render_frame_host &&
+      render_frame_host_id ==
+          web_contents()->GetPrimaryMainFrame()->GetGlobalId()) {
+    GetRemote(render_frame_host)
+        ->RequestAsyncExecuteScript(world_id, base::UTF8ToUTF16(script), false,
+                                    true, true, std::move(cb));
+  } else {
+    VLOG(2) << "render_frame_host is invalid.";
+    return;
+  }
+}
+
+mojo::AssociatedRemote<script_injector::mojom::ScriptInjector>&
+PsstTabHelper::GetRemote(content::RenderFrameHost* rfh) {
+  if (!script_injector_remote_.is_bound()) {
+    rfh->GetRemoteAssociatedInterfaces()->GetInterface(
+        &script_injector_remote_);
+  }
+  return script_injector_remote_;
+}
+
+void PsstTabHelper::DidFinishNavigation(
+    content::NavigationHandle* navigation_handle) {
+  if (!navigation_handle->IsInPrimaryMainFrame() ||
+      !navigation_handle->HasCommitted() ||
+      navigation_handle->IsSameDocument()) {
+    return;
+  }
+
+  DCHECK(psst_rule_service_);
```

Comment:
this seems out of place, you don't even access `psst_rule_service_` here. I think you only need this check in the constructor